xaitk_saliency/impls/gen_object_detector_blackbox_sal/occlusion_based.py

```diff
@@ -0,0 +1,172 @@
+    `PerturbImage` implementation.
+    This is a parameter to be set during runtime as this is most often driven
+    by the black-box algorithm used, if at all.
+
+    :param perturber: `PerturbImage` implementation instance for generating
+        occlusion masks.
+    :param generator: `GenerateDetectorProposalSaliency` implementation
+        instance for generating saliency masks given occlusion masks and
+        black-box detector outputs.
+    :param threads: Optional number threads to use to enable parallelism in
+        applying perturbation masks to an input image. If 0, a negative value,
+        or `None`, work will be performed on the main-thread in-line.
+    """
+
+    def __init__(
+        self,
+        perturber: PerturbImage,
+        generator: GenerateDetectorProposalSaliency,
+        threads: Optional[int] = 0
+    ):
+        self._perturber = perturber
+        self._generator = generator
+        self._threads = threads
+        # Optional fill color
+        self.fill: Optional[Union[int, Sequence[int]]] = None
+
+    def _generate(
+        self,
+        ref_image: np.ndarray,
+        bboxes: np.ndarray,
+        scores: np.ndarray,
+        blackbox: DetectImageObjects,
+        objectness: Optional[np.ndarray] = None,
+    ) -> np.ndarray:
+        ref_dets_mat = format_detection(bboxes, scores, objectness)
+
+        pert_masks = self._perturber(ref_image)
+
+        pert_dets = blackbox.detect_objects(
+            occlude_image_streaming(
+                ref_image,
+                pert_masks,
+                fill=self.fill,
+                threads=self._threads
+            )
+        )
+
+        pert_dets_mat = _dets_to_formatted_mat(pert_dets)
+
+        return self._generator(
+            ref_dets_mat,
+            pert_dets_mat,
+            pert_masks
+        )
+
+    @classmethod
+    def get_default_config(cls) -> Dict[str, Any]:
+        cfg = super().get_default_config()
+        cfg['perturber'] = make_default_config(PerturbImage.get_impls())
+        cfg['generator'] = make_default_config(GenerateDetectorProposalSaliency.get_impls())
+        return cfg
+
+    @classmethod
+    def from_config(
+        cls: Type[C],
+        config_dict: Dict,
+        merge_default: bool = True
+    ) -> C:
+        config_dict = dict(config_dict)  # shallow-copy
+        config_dict['perturber'] = from_config_dict(
+            config_dict['perturber'],
+            PerturbImage.get_impls()
+        )
+        config_dict['generator'] = from_config_dict(
+            config_dict['generator'],
+            GenerateDetectorProposalSaliency.get_impls()
+        )
+        return super().from_config(config_dict, merge_default=merge_default)
+
+    def get_config(self) -> Dict[str, Any]:
+        return {
+            "perturber": to_config_dict(self._perturber),
+            "generator": to_config_dict(self._generator),
+            "threads": self._threads,
+        }
+
+
+def _dets_to_formatted_mat(
```

Comment:
There is a similar function to this in `xaitk_saliency/utils/gen_coco_sal.py`. Is there any deduplication that we can do with that, e.g. a common function in `xaitk_saliency/utils/`?

Reply:
The function in `gen_coco_sal` is a bit different, it uses the categories from the `kwcoco.CocoDataset` to maintain constant score ordering. I think it might make sense to convert both of these to a common function, but that would require changing the logic in `gen_coco_sal`. Since we are discussing getting rid of or changing `gen_coco_sal`, I think that can wait.